funcr/funcr.go

```diff
@@ -168,12 +168,20 @@ const (
 
 // TODO: This is not fast. Most of the overhead goes here.
 func prettyWithFlags(value interface{}, flags uint32) string {
-	// Handling the most common types without reflect is a small perf win.
-	switch v := value.(type) {
-	case logr.Marshaler:
+	// Handle fundamental interface tests before we start processing the
+	// value.
+	if v, ok := value.(logr.Marshaler); ok {
 		// Replace the value with what the value wants to get logged.
 		// That then gets handled below via reflection.
 		value = v.MarshalLog()
+	}
+	if v, ok := value.(fmt.Stringer); ok {
+		// Handle types that want to format themselves.
+		value = v.String()
+	}
```

Comment:
Should we also check for the `error` interface here? The justification is similar as for `String`:
- some error types have no exported fields and logging them via reflection is useless
- some error types contain so many fields that logs explode (I've seen that for gRPC errors in Ginkgo assertions)

Reply:
I thought about that too, just wanted to float this first